server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceClusterStateDelay.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices.recovery;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.SubscribableListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+public class PeerRecoverySourceClusterStateDelay {
+    private PeerRecoverySourceClusterStateDelay() {}
+
+    private static final Logger logger = LogManager.getLogger(PeerRecoverySourceClusterStateDelay.class);
+
+    /**
+     * Waits for the given cluster state version to be applied locally before proceeding with recovery
+     */
+    public static <T> void ensureClusterStateVersion(
```

Comment:
This is the place that implements the waiting. Extracted to a utility class so we can use it in Serverless too.